2.2-taptweak.ipynb

```diff
@@ -372,6 +443,262 @@
     "test.shutdown()"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Part 3 (Case Study): Contract Commitments\n",
+    "\n",
+    "Alice currently commits contracts with Bob to unspendable OP_RETURN outputs, which contain 32B proof-of-existence commitments. Although this is a standard output with a zero amount, several disadvantages remain.\n",
+    "\n",
+    "* Committing data to an OP_RETURN output requires an additional output with a zero amount, resulting in a higher transaction fees.\n",
+    "* The OP_RETURN output reveals the presence of a data commitment to any on-chain observer. This reduces the privacy of Alice's commitments.\n",
+    "\n",
+    "In this chapter, we'll show how Alice can move her contract commitments to public key tweaks to reduce fees and improve the privacy of her commitments."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Committing contract data to an OP_RETURN output\n",
+    "\n",
+    "We'll first show Alice's current setup: An OP_RETURN script containing commitment data."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Example 2.2.9: Create the contract commitment"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "contract_bytes = \"Alice pays 10 BTC to Bob\".encode('utf-8')\n",
+    "commitment_bytes = sha256(contract_bytes)\n",
+    "print(\"The contract commitment is: {}\".format(commitment_bytes.hex()))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Example 2.2.10: Create and broadcast a transaction with an OP_RETURN output\n",
+    "\n",
+    "We now construct a OP_RETURN output which contains the commitment data of Alice's contract with Bob, and then add it to a transaction with a regular pay-to-pubkey output. This way, the commitment can be done more efficiently, by sharing transaction data with another spendable output."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Start node\n",
+    "test = util.TestWrapper()\n",
+    "test.setup()\n",
+    "node = test.nodes[0]\n",
+    "\n",
+    "# Generate coins to a new wallet address\n",
+    "coinbase_address = node.getnewaddress(address_type=\"bech32\")\n",
+    "node.generatetoaddress(101, coinbase_address)\n",
+    "balance = node.getbalance()\n",
+    "assert balance > 1\n",
```

Comment:
Perhaps `generate_and_send_coins()` here. Also not sure we need the balance var or assert statement

Reply:
Ok, I will open a separate PR to modify `generate_and_send_coins()` to accept multiple outputs and both addresses and data (op_return).

Reply:
Ok, this is now tracked here #143.